### PR TITLE
Unions of PlutusData and BuiltinData

### PR DESCRIPTION
**Pull request by SCMusson**
Targeting issue #367 
The following would now compile
```
source_code = """
from dataclasses import dataclass
from typing import Dict, List, Union
from pycardano import Datum as Anything, PlutusData

@dataclass()
class A(PlutusData):
    CONSTR_ID = 0
    foo: int

def validator(x: Union[A, int, bytes, List[int], Dict[int, int]]) -> int:
    k: int = 0
    if isinstance(x, A):
        k = 5
    elif isinstance(x, bytes):
        k = 7
    elif isinstance(x, int):
        k = 6
    elif isinstance(x, List[int]):
        k = 8
    elif isinstance(x, Dict[int, int]):
        k = 9
    return k
"""
```

I think it can't currently distinguish between `Dict[int, int]` and `Dict[int, bytes]` for example. I'll work on that if you think this is heading in the correct direction. I also need to sit down and think of what edge cases might break this and write some more tests.

### CODE REVIEW

**Comment by nielstron** [comment]
I think it would be best to only allow joining Dict[Any, Any] and List[Any] in these unions, ideally making dict and list synonyms of this type.

If a user really knows what they are doing maybe an unsafe cast from dict to Dict[int, int] can then be allowed?

Also it's important that str is not allowed in these unions but this seems to be the case already.

What happens when a function has parameter type Union[A, int] and is passed an int/A (where the type of that variable was just int / A before)? Need to wrap in PlutusData in those cases

File: opshin/fun_impls.py [hunk 1]
```diff
@@ -87,6 +87,80 @@ def impl_from_args(self, args: typing.List[Type]) -> plt.AST:
...
+                    plt.Bool(False),
+                    plt.Bool(True),
+                ),
+            )
+        elif isinstance(args[1], RecordType):
+            return OLambda(
+                ["x"],
+                plt.ChooseData(
+                    OVar("x"),
+                    plt.Bool(True),
```
[the remark this answers is not in the document]
Comment (SCMusson) [reply opshin/fun_impls.py:127]: If there are both builtin and plutusData types in a union, then when checking plutusData it first checks its a record type and then invokes the old functionality of checking CONSTR_ID:
https://github.com/OpShin/opshin/blob/e0e5d25b7702c2f930f5719a5cf5826af4d01d46/opshin/type_inference.py#L962-L981

File: opshin/type_inference.py [hunk 2]
```diff
@@ -705,10 +709,15 @@ def visit_FunctionDef(self, node: FunctionDef) -> TypedFunctionDef:
...
         tfd.args = self.visit(node.args)
+
         functyp = FunctionType(
             frozenlist([t.typ for t in tfd.args.args]),
             InstanceType(self.type_from_annotation(tfd.returns)),
-            bound_vars={v: self.variable_type(v) for v in externally_bound_vars(node)},
+            bound_vars={
+                v: self.variable_type(v)
+                for v in externally_bound_vars(node)
+                if not v in ["List", "Dict"]
```
[the remark this answers is not in the document]
Comment (SCMusson) [reply opshin/type_inference.py:719]: Neither List or Dict appear in `self.scopes` (so self.variable_type will fail) but end up in `externally_bound_vars`.  I'm not sure what the most 'correct' solution here is. I'm guessing there would be problems if someone tried to define there own class with the name `Test` and `Dict`.

**Comment by SCMusson** [comment]
I've just realised too that `isinstance(x, List[int])` and `isinstance(x, Dict[int, int])` is not valid python.
The following would be thrown:

> TypeError: Subscripted generics cannot be used with class and instance checks


So would the following be what you are looking for:
 perhaps using `isinstance(x, list)` and `inisntance(x, dict)`.
```
source_code = """
from dataclasses import dataclass
from typing import Dict, List, Union
from pycardano import Datum as Anything, PlutusData

@dataclass()
class A(PlutusData):
    CONSTR_ID = 0
    foo: int

def validator(x: Union[A, int, bytes, List[Anything], Dict[Anything, Anything]]) -> int:
    k: int = 0
    if isinstance(x, A):
        k = 5
    elif isinstance(x, bytes):
        k = 7
    elif isinstance(x, int):
        k = 6
    elif isinstance(x, List):
        k = 8
    elif isinstance(x, Dict):
        k = 9
    return k
"""
``` 

Alternatively working towards using `isinstance(x, list)` and `isinstance(x, dict)` instead or as well as `isinstanec(x, List)` and `isinstance(x, Dict)`.

Let me know what you think.

**Comment by nielstron** [comment]
Nice catch! If isinstance(x, Dict) is valid python that should be fine, otherwise isinstance(x, dict) would be required.

**Comment by SCMusson** [comment]
I've had some free time so I've come back to this.
The following should run:
```
source_code = """
from dataclasses import dataclass
from typing import Dict, List, Union
from pycardano import Datum as Anything, PlutusData

@dataclass()
class A(PlutusData):
    CONSTR_ID = 0
    foo: int

def validator(x: Union[A, int, bytes, List[Anything], Dict[Anything, Anything]]) -> int:
    k: int = 0
    if isinstance(x, A):
        k = 5
    elif isinstance(x, bytes):
        k = 7
    elif isinstance(x, int):
        k = 6
    elif isinstance(x, List):
        k = 8
    elif isinstance(x, Dict):
        k = 9
    return k
"""
```
Only List[Anything] and Dict[Anything, Anything] are accepted in a Union now.
It will raise an error if List or Dict are subscripts inside isinstance to align with actual python.
Looks like python 3.11 CI errored because of timings not because of bugs from what I'm reading.

File: opshin/tests/test_Unions.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+
+def to_int(x):
+    if isinstance(x, A):
+        return 5
+    elif isinstance(x, int):
+        return 6
+    elif isinstance(x, bytes):
+        return 7
+    elif isinstance(x, list):
```
Comment (nielstron) [inline opshin/tests/test_Unions.py:22]: In order to make really sure that Python and OpShin map to the same operaton here (i.e. that List is allowed for isinstance)
```suggestion
    elif isinstance(x, List):
```

File: opshin/tests/test_Unions.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        return 8
+    elif isinstance(x, dict):
```
Comment (nielstron) [inline opshin/tests/test_Unions.py:24]:
```suggestion
    elif isinstance(x, Dict):
```

File: opshin/tests/test_Unions.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        return 9
+    return False
+
+
+union_types = st.sampled_from([A(0), 10, b"foo", [1, 2, 3, 4, 5], {1: 2, 2: 3}])
+
+
+class Union_tests(unittest.TestCase):
```
Comment (nielstron) [inline opshin/tests/test_Unions.py:32]: Please add a testcase that shows that Unioning PlutusData of the same constructors is illegal. Also one for str, bool (the non-allowed primitive type).

**Review by nielstron — CHANGES_REQUESTED** [review]
Looks great now. Just some small comments.

**Comment by SCMusson** [comment]
Implemented those requested changes. 
Also test for failure with Bool, Str, and duplicate constr_ids in Unions. 
I added a check for duplicate constri_ids in Unions.

**Review by nielstron — APPROVED** [review]

**Comment by nielstron** [comment]
As we have discovered a few missing edge cases here (see #403) I will not be able to pay out the full bug bounty for this PR. If @SCMusson is willing to supply the additional fixes I would simply offer additional time to resolve this and claim the full reward, otherwise we will have to decide a reasonable split with another developer.

**Comment by SCMusson** [comment]
I plan to return to this problem, however I will be short of available time for the next week after which I will be able to devote some time to fixing this.

**Comment by nielstron** [comment]
Sure, no problem 👍
